File: Data Structures & Algorithms/trapping-rain-water/submission-4.java
class Solution {
    public int trap(int[] height) {
        int left = 0;
        int right = 1;
        int res = 0;
        
        int[] water = new int[height.length];

        while (left < height.length && right < height.length) {
            if (height[right] < height[left]) {
                water[right] = height[left] - height[right];
                right++;
            }
            else {
                left = right;
                right++;
            }
        }

        left = height.length - 2;
        right = height.length - 1;

        while (left > 0 && right > 0) {
            if (height[right] > height[left]) {
                res += Math.min(water[left], height[right] - height[left]);
                left--;
            }
            else {
                right = left;
                left--;
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []

        def dfs(curr, res, height):
            if not curr:
                return
            if height == len(res):
                res.append([curr.val])
            else:
                res[height].append(curr.val)
            dfs(curr.left, res, height + 1)
            dfs(curr.right, res, height + 1)

        dfs(root, res, 0)
        return res

        
